mantidproject/mantid
# synchronized ADS event handling in WorkspaceSelector

**Description of work.**

It has been found that relatively rapid changes in the ADS contents can cause hard crash coming from the `WorkspaceSelector`. The issue is revealed as the new `WorkspaceCalculator` widget has 3 selectors and is docked, so it is present all the time. The main issue was that the `WorkspaceSelector` observes the ADS and the handlers were not thread protected. If a notification comes while handling another one, the structure of the underlying QComboBox is not consistent. 
 
The proposed solution is as follows:
1. Make ADS event handlers in the `WorkspaceSelector` synchronized.
2. In `WorkspaceCalculator` :
    a) Do not observe ADS if the widget is hidden, this is for extra security.
    b) `WorkspaceSelector`s are set to not observe ADS at the construction time
    c) They are set to observe on`focusInEvent` 
    d) They disconnect on `activated` signal

There are still two scenarios to cause a crash, which will be investigated separately.
This is when one opens a selector but doesn't make a choice and keeps it open. This means that focusIn is triggered, but activate is not, so the widget stays in observing state. The other is when a selector is opened while the script is running.
However, this PR already reduces the chance of the crash to a minimum, so it should get in regardless.

**To test:**
The following script exercises the issue:
Note that it creates numerous small (single-valued) workspaces in a script; a scenario that induces the crash.
With real-sized workspaces this doesn't occur as the creation/manipulation time of the workspaces gives enough time for the widget to react.

```
from mantid.simpleapi import *
ws_list = []

for i in range(50000):
    print(i)
    name = "ws" + str(i)
    CreateSingleValuedWorkspace(OutputWorkspace=name, DataValue=0)
    ws_list.append(name)
    if i % 10 == 0:     
        DeleteWorkspaces(WorkspaceList=ws_list)
        ws_list.clear()
```

Case 1:
1. Hide WorkspaceCalculator using the menu
2. Run the script, make sure there is no crash

Case 2:
1. Show the WorkspaceCalculator, but don't touch it
2. Run the script, make sure there is no crash

Case 3:
1. Run the script, and interrupt at some point
2. Click around on the widget, choose some workspaces in the selectors
3. Run the script, make sure there is no crash

Case 4 (still crashing):
1. Click on one of the selectors, but do not choose anything
2. Run the script while the list is open (you will need to do Ctrl+Enter to avoid clicking)

Case 5 (still crashing):
1. Run the script
2. Click on one of the selectors while the script is running

Ideas are welcome for 4 and 5, but this is a type of problem that is hard to track down, since the crash doesn't happen on debug build. 

Fixes #32032 . 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
